contracts/external/cw-payroll-factory/src/contract.rs

```diff
@@ -0,0 +1,325 @@
+            recipient,
+            start_before,
+            limit,
+        } => {
+            let limit = limit.unwrap_or(DEFAULT_LIMIT).min(MAX_LIMIT) as usize;
+            let start = start_before.map(Bound::<String>::exclusive);
+
+            // Validate recipient address
+            deps.api.addr_validate(&recipient)?;
+
+            let res: Vec<VestingContract> = vesting_contracts()
+                .idx
+                .recipient
+                .prefix(recipient)
+                .range(deps.storage, None, start, Order::Descending)
+                .take(limit)
+                .flat_map(|vc| Ok::<VestingContract, ContractError>(vc?.1))
+                .collect();
+
+            Ok(to_binary(&res)?)
+        }
+        QueryMsg::Ownership {} => to_binary(&cw_ownable::get_ownership(deps.storage)?),
+    }
+}
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn reply(deps: DepsMut, _env: Env, msg: Reply) -> Result<Response, ContractError> {
+    match msg.id {
+        INSTANTIATE_CONTRACT_REPLY_ID => {
+            let res = parse_reply_instantiate_data(msg)?;
+            let contract_addr = deps.api.addr_validate(&res.contract_address)?;
+
+            // Query new vesting payment contract for info
+            let vesting_payment: VestingPayment = deps
+                .querier
+                .query_wasm_smart(contract_addr.clone(), &PayrollQueryMsg::Info {})?;
+
+            let instantiator = TMP_INSTANTIATOR_INFO.load(deps.storage)?;
+
+            // Save vesting contract payment info
+            vesting_contracts().save(
+                deps.storage,
+                contract_addr.as_ref(),
+                &VestingContract {
+                    instantiator: instantiator.to_string(),
+                    recipient: vesting_payment.recipient.to_string(),
+                    contract: contract_addr.to_string(),
+                },
+            )?;
+
+            // Clear tmp instatiator info
+            TMP_INSTANTIATOR_INFO.remove(deps.storage);
+
+            // If cw20, fire off fund message!
+            let msgs: Vec<CosmosMsg> = match vesting_payment.denom {
+                CheckedDenom::Native(_) => vec![],
+                CheckedDenom::Cw20(denom) => {
+                    // Send transaction to fund contract
+                    vec![CosmosMsg::Wasm(WasmMsg::Execute {
+                        contract_addr: denom.to_string(),
+                        msg: to_binary(&Cw20ExecuteMsg::Send {
+                            contract: contract_addr.to_string(),
+                            amount: vesting_payment.amount,
+                            msg: to_binary(&PayrollReceiveMsg::Fund {})?,
+                        })?,
+                        funds: vec![],
+                    })]
+                }
+            };
+
+            Ok(Response::default()
+                .add_attribute("new_payroll_contract", contract_addr)
+                .add_messages(msgs))
+        }
```

Comment:
Do I understand correctly that you could move all this code to `instantiate_contract` if you knew the address of the new instance in advance, avoiding the need for `TMP_INSTANTIATOR_INFO`?

I think Instantiate2 coming in CosmWasm 1.2 will give you that.

Reply:
Yes. That would work.